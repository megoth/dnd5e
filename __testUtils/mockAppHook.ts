import mockApp from "./mockApp";

export default function mockAppHook(mock: jest.Mock, app = mockApp()) {
  mock.mockReturnValue(app);
}
