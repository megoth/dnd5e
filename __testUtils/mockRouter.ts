import { NextRouter } from "next/router";

export default function mockRouter(
  overrides: Partial<NextRouter> = {}
): NextRouter {
  return {
    asPath: "/",
    query: {},
    ...overrides,
  } as NextRouter;
}
