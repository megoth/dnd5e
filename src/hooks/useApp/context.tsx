import { createContext } from "react";
import { Locale } from "../../ldo/app.typings";

const AppContext = createContext<{
  currentLocale: string;
  availableLocales: Array<Locale>;
}>({
  currentLocale: "en-US",
  availableLocales: [],
});

export default AppContext;
