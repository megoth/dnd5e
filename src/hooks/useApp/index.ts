import { useContext, useEffect } from "react";
import { AppContext } from "../../contexts/app";
import { appIsLoadingLocalizations } from "../../models/app";

export default function useApp(bundles: string[] = []) {
  const { app, setBundles } = useContext(AppContext);
  useEffect(() => {
    if (!app || !appIsLoadingLocalizations(app, bundles)) {
      return;
    }
    setBundles(bundles);
  }, [app, bundles, setBundles]);
  return app;
}
