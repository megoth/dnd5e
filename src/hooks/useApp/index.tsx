import { useContext } from "react";
import AppContext from "./context";

export default function useApp() {
  return useContext(AppContext);
}
