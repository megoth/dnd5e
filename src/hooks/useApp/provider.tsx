import React, { ReactNode, useEffect, useMemo, useState } from "react";
import { LocalizationProvider, ReactLocalization } from "@fluent/react";
import { getLocale } from "../../utils/language";
import AppContext from "./context";
import useSWR from "swr";
import Loading from "../../components/loading";
import {
  AppShapeType,
  LocaleShapeType,
  ResourceBundleShapeType,
  TranslationShapeType,
  TranslationsIndexShapeType,
} from "../../ldo/app.shapeTypes";
import { SessionInfo, useLdo, useSolidAuth } from "@ldo/solid-react";
import { hash, resourceUrl } from "../../utils/url";
import { useSearchParams } from "react-router-dom";
import { FluentBundle, FluentResource } from "@fluent/bundle";
import { userIsAdmin } from "../../utils/session";
import { namedNode } from "@rdfjs/data-model";
import { useLocalStorage } from "@uidotdev/usehooks";
import { Literal } from "@rdfjs/types";

interface Props {
  children: ReactNode;
}

function getBundles(session: SessionInfo): string[] {
  return userIsAdmin(session) ? ["global", "admin"] : ["global"];
}

export default function AppProvider({ children }: Props) {
  const { session } = useSolidAuth();
  const { dataset, getResource, getSubject } = useLdo();
  const [searchParams] = useSearchParams();
  const [currentLocale, setCurrentLocale] = useLocalStorage("locale", "en-US");
  const [bundles, setBundles] = useState<string[]>(getBundles(session));

  useEffect(() => {
    setBundles(getBundles(session));
  }, [session]);

  // fetch app index
  const { data: app } = useSWR("app", async () => {
    const appUrl =
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      import.meta.env.VITE_APP_URL ||
      new URL("/data/index.ttl#dnd5e", location.origin).toString();
    const path = resourceUrl(appUrl);
    await getResource(path).readIfUnfetched();
    return getSubject(AppShapeType, appUrl);
  });

  // fetch resource bundles
  const { data: resourceBundles } = useSWR(
    () => `resourceBundles-${app["@id"]}}-${bundles.join("-")}`,
    async () =>
      (
        await Promise.all(
          app.resourceBundle.map(async (bundle) => {
            await getResource(bundle["@id"]).readIfUnfetched();
            return getSubject(ResourceBundleShapeType, bundle["@id"]);
          }),
        )
      ).filter((bundle) => bundles.indexOf(bundle.label) !== -1),
  );

  // fetch supported languages
  const { data: supportedLocales } = useSWR(
    () => `supportedLocales-${app["@id"]}}`,
    async () => {
      return Promise.all(
        app.supportLanguage.map(async (language) => {
          await getResource(language["@id"]).readIfUnfetched();
          return getSubject(LocaleShapeType, language["@id"]);
        }),
      );
    },
  );

  useEffect(() => {
    if (!supportedLocales) return;
    // TODO: FIX so that you can load translations directly
    const requestedLocale = searchParams.get("locale");
    const locale = getLocale(
      (Array.isArray(requestedLocale) ? requestedLocale[0] : requestedLocale) ||
        currentLocale,
      supportedLocales?.map((locale) => locale.language),
    );
    setCurrentLocale(locale);
  }, [searchParams.get("locale"), supportedLocales, currentLocale]);

  // fetch FAQs
  const { isLoading: faqIsLoading } = useSWR(
    () => `faqs-${resourceBundles.map((b) => b["@id"]).join("-")}}`,
    async () =>
      Promise.all(
        resourceBundles
          .flatMap((bundle) => bundle.faqIndex)
          .map(async (index) => getResource(index["@id"]).readIfUnfetched()),
      ),
  );

  // fetch translation indices
  const { data: translations } = useSWR(
    () =>
      `translationsIndices-${resourceBundles.map((b) => b["@id"]).join("-")}-${currentLocale}`,
    async () => {
      return (
        await Promise.all(
          resourceBundles
            .flatMap((bundle) => bundle.translationsIndex)
            .map(async (index) => {
              await getResource(index["@id"]).readIfUnfetched();
              return getSubject(TranslationsIndexShapeType, index["@id"]);
            }),
        )
      ).filter((index) => !index.language || index.language === currentLocale);
    },
  );

  // get translations
  const { isLoading: translationsLoading } = useSWR(
    () => `translations-${translations.map((t) => t["@id"]).join("-")}}`,
    async () =>
      Promise.all(
        translations
          .map((index) => index.translationsResource)
          .map(async (resourceUrl) => {
            return getResource(resourceUrl["@id"]).readIfUnfetched();
          }),
      ),
  );
  const localization = useMemo(() => {
    const localization = new FluentBundle(currentLocale);
    if (!resourceBundles || translationsLoading) return null;
    resourceBundles.forEach((bundle) =>
      bundle.translationsIndex
        .map((index) => index.translationsResource["@id"])
        .flatMap((resourceUrl) => {
          return dataset
            .match(null, null, null, namedNode(resourceUrl))
            .toArray()
            .filter(
              (quad) =>
                (quad.object as Literal).language ===
                currentLocale.toLowerCase(),
            )
            .map((quad) =>
              getSubject(TranslationShapeType, quad.subject.value),
            );
        })
        .forEach((translation) =>
          localization.addResource(
            new FluentResource(
              `${hash(translation["@id"])} = ${translation.definition}`,
            ),
          ),
        ),
    );
    return new ReactLocalization([localization]);
  }, [currentLocale, resourceBundles, dataset, translationsLoading]);

  if (translationsLoading || faqIsLoading || !translations || !localization) {
    return <Loading />;
  }

  return (
    <AppContext.Provider
      value={{
        app,
        currentLocale,
        availableLocales: supportedLocales?.filter(
          (locale) => locale.language !== currentLocale,
        ),
      }}
    >
      <LocalizationProvider l10n={localization}>
        {children}
      </LocalizationProvider>
    </AppContext.Provider>
  );
}
