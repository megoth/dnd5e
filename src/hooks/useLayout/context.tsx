import { createContext } from "react";

export type SetBooleanValue = (menuOpen: boolean) => void;

const LayoutContext = createContext<{
  full: boolean;
  leftOpen: boolean;
  rightOpen: boolean;
  setFull: SetBooleanValue;
  setLeftOpen: SetBooleanValue;
  setRightOpen: SetBooleanValue;
}>({
  full: false,
  leftOpen: false,
  rightOpen: false,
  setLeftOpen: () => {},
  setFull: () => {},
  setRightOpen: () => {},
});

export default LayoutContext;
