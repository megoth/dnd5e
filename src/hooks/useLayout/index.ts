import { useContext } from "react";
import { LayoutContext } from "../../contexts/layout";

export default function useLayout() {
  return useContext(LayoutContext);
}
