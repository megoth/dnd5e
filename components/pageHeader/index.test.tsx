import React from "react";
import { render } from "@testing-library/react";
import * as routerFns from "next/router";
import userEvent from "@testing-library/user-event";
import * as solidUIReactFns from "@inrupt/solid-ui-react";
import useApp from "../../src/hooks/useApp";
import mockAppHook from "../../__testUtils/mockAppHook";
import PageHeader, {
  TESTID_PAGE_HEADER_LEFT_MENU_BUTTON,
  TESTID_PAGE_HEADER_RIGHT_MENU_BUTTON,
} from "./index";
import useLayout from "../../src/hooks/useLayout";
import {
  mockAuthenticatedSession,
  mockUnauthenticatedSession,
} from "../../__testUtils/mockSession";
import mockRouter from "../../__testUtils/mockRouter";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

jest.mock("../../src/hooks/useLayout");
const mockedLayoutHook = useLayout as jest.Mock;

describe("PageHeader", () => {
  const authenticatedSession = mockAuthenticatedSession();

  let setLeftOpen;
  let setRightOpen;
  let mockedSessionHook;
  let mockedRouterHook;

  beforeEach(() => mockAppHook(mockedAppHook));
  beforeEach(() => {
    jest
      .spyOn(routerFns, "useRouter")
      .mockReturnValue(mockRouter({ asPath: "/" }));
  });
  beforeEach(() => {
    setLeftOpen = jest.fn();
    setRightOpen = jest.fn();
    mockedLayoutHook.mockReturnValue({ setLeftOpen, setRightOpen });
  });
  beforeEach(() => {
    mockedSessionHook = jest
      .spyOn(solidUIReactFns, "useSession")
      .mockReturnValue(authenticatedSession);
  });
  beforeEach(() => {
    mockedRouterHook = jest.spyOn(routerFns, "useRouter");
  });

  it("renders", () => {
    const { asFragment } = render(<PageHeader />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders button that opens the right menu", () => {
    const { getByTestId } = render(<PageHeader />);
    userEvent.click(getByTestId(TESTID_PAGE_HEADER_RIGHT_MENU_BUTTON));
    expect(setRightOpen).toHaveBeenCalledWith(true);
  });

  it("renders different icon when unauthenticated", () => {
    mockedSessionHook.mockReturnValue(mockUnauthenticatedSession());
    const { asFragment } = render(<PageHeader />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders button that opens the left menu", () => {
    mockedRouterHook.mockReturnValue({ asPath: "/href/test" });
    const { getByTestId } = render(<PageHeader />);
    userEvent.click(getByTestId(TESTID_PAGE_HEADER_LEFT_MENU_BUTTON));
    expect(setLeftOpen).toHaveBeenCalledWith(true);
  });
});
