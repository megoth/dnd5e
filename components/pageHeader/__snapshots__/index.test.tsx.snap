// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader renders 1`] = `
<DocumentFragment>
  <header
    class="mb-2 shadow shadow-md"
  >
    <div
      class="main-container main-container--content"
    >
      <div
        class="font-serif self-center text-left"
      >
        <a
          class="focus:outline-none focus:ring-2 focus:ring-red-600"
          href="/"
        >
          [Translation for https://example.com/global/translations.ttl#appName does not exist]
        </a>
      </div>
    </div>
    <nav
      class="pages-nav pages-nav--main-menu"
    >
      <ul
        class="pages-nav__list pages-nav__list--main-menu"
      >
        <li
          class="flex"
          data-testid="main-nav-list-item"
        >
          <a
            class="pages-nav__link pages-nav__link--main-menu"
            href="/characters"
          >
            [Translation for https://example.com/global/translations.ttl#charactersPageTitle does not exist]
          </a>
        </li>
        <li
          class="flex"
          data-testid="main-nav-list-item"
        >
          <a
            class="pages-nav__link pages-nav__link--main-menu"
            href="/rules"
          >
            [Translation for https://example.com/global/translations.ttl#rulesPageTitle does not exist]
          </a>
        </li>
      </ul>
    </nav>
  </header>
</DocumentFragment>
`;
